### PR TITLE
DPoP documentation

### LINKED ISSUE #21917: DPoP documentation
### Description

Docs for DPoP, which was implemented in https://github.com/keycloak/keycloak/issues/21200

### PR DESCRIPTION
**Pull request by tnorimat**
closes #21917

### CODE REVIEW

**Comment by tnorimat** [comment]
@mposolda The PR is for a follow-up issue of https://github.com/keycloak/keycloak/issues/21917 . Could you check it?

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc [hunk 1]
```diff
@@ -94,6 +94,39 @@ See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-mtls-08#section-3[Mut
...
+. {project_name} requests a DPoP proof.
+. {project_name} receives the DPoP proof.
+. {project_name} successfully verifies the DPoP proof.
+
+If verification fails, {project_name} rejects the token.
+
+In the following cases, {project_name} will verify the client sending the access token or the refresh token:
+
+* A token refresh request is sent to the token endpoint with a holder-of-key refresh token.
+* A logout request is sent to Logout endpoint with a holder-of-key refresh token.
```
Comment (mposolda) [inline docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]:
```suggestion
* A logout request is sent to Logout endpoint with a holder-of-key refresh token. This is done only for public clients as described above.
```

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat I've added 2 comments inline. Could you please check them?

Besides that, I wonder if we should document also the things from the follow-up tasks #21919 and #21920 and there might be some changes to the documentation needed based on those tasks?

Specifically I think the additional things to document based on those might be:
- We can document that if the switch is ON, the DPoP binding is enforced. If the switch is OFF, the DPoP binding is still possible if `DPoP` header is present in the token request, but it is not enforced (This is related to #21920)
- We can add something about support of token-introspection endpoint and possibility to use `DPoP` as `token_type` as described in the specification (related to #21919)

@tnorimat I don't know if we want to add these documentation now or if we add them during follow-up PR on the individual tasks? Whatever you prefer.

The advantage of doing all the documentation in this PR is that we won't need to ask documentation team (Andrew Munro) for the review of the follow-up PRs. But at the same time, the documentation won't be strictly accurate to what is currently in the main. So I don't have clear preference on whether to do all the docs now or whether to change docs in the follow-up PRs. I am fine with what you prefer. WDYT?

**Comment by tnorimat** [comment]
@mposolda Thank you for your comments. I will update the documantation as you suggested. And I agree with the idea that we should add the documentation after resolving https://github.com/keycloak/keycloak/issues/21919 and https://github.com/keycloak/keycloak/issues/21920 . I would like to do that.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 1 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]: Sorry, the DPoP 1st step PR does not include DPoP proof and bind verification on Logout endpoint. Therefore, I would like to delete the sentence.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc [hunk 2]
```diff
@@ -94,6 +94,39 @@ See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-mtls-08#section-3[Mut
...
+. A token request is sent to the token endpoint in an authorization code flow or hybrid flow.
+. {project_name} requests a DPoP proof.
+. {project_name} receives the DPoP proof.
+. {project_name} successfully verifies the DPoP proof.
+
+If verification fails, {project_name} rejects the token.
+
+In the following cases, {project_name} will verify the client sending the access token or the refresh token:
+
+* A token refresh request is sent to the token endpoint with a holder-of-key refresh token.
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:115]: I will add the sentence.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 1 above)
Comment (mposolda) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc]: Well, It's not too obvious, but I think you were correct and it actually does also for logout endpoint :-) Method `LogoutEndpoint.logoutToken()` calls `tokenManager.verifyRefreshToken`, which eventually validates DPoP binding. 

So I vote to add it back, maybe with something like:
```
* A logout request is sent to Logout endpoint with a holder-of-key refresh token. This is done only for public clients as described above.
```

What do you think?

Comment (mposolda) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc]: By the way, I think you know that this is Keycloak proprietary logout endpoint (not OIDC based) with refreshToken sent as an argument. This endpoint is used by our (deprecated) java adapters. Not sure if to emphasize it in the documentation as well?

In this case, we should possibly also update the corresponding paragraph in the MTLS HoK section above. But if you prefer to not emphasize it, it is fine to me.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc]: I added the sentence.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 1 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc]: Yes, I restored the sentence for Logout Endpoint.
Also, I added "non-OIDC keycloak proprietary Logout Endpoint" to both descriptions for OAuth-MTLS and DPoP.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc [hunk 3]
```diff
@@ -94,6 +94,39 @@ See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-mtls-08#section-3[Mut
...
+. {project_name} requests a DPoP proof.
+. {project_name} receives the DPoP proof.
+. {project_name} successfully verifies the DPoP proof.
+
+If verification fails, {project_name} rejects the token.
+
+In the following cases, {project_name} will verify the client sending the the refresh token:
+
+* A token refresh request is sent to the token endpoint with a holder-of-key refresh token. This verification is done only for public clients as described in the DPoP specification. For confidential clients, the verification is not done as client authentication with proper client credentials is in place to ensure that request comes from the legitimate client. For public clients, both access tokens and refresh tokens are DPoP bound. For confidential clients only access tokens are DPoP bound.
+* A logout request is sent to non-OIDC compliant {project_name} proprietary logout endpoint Logout endpoint with a holder-of-key refresh token. This is done only for public clients as described above.
```
Comment (mposolda) [inline docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]:
```suggestion
* A logout request is sent to non-OIDC compliant {project_name} proprietary Logout endpoint with a holder-of-key refresh token. This is done only for public clients as described above.
```

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat I've added 1 comment inline (very small change with duplicated wording added in the sentence)

Also I am adding @andymunro from our documentation team as a reviewer.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc [hunk 4]
```diff
@@ -85,7 +85,7 @@ In the following cases, {project_name} will verify the client sending the access
 
 * A token refresh request is sent to the token endpoint with a holder-of-key refresh token.
 * A UserInfo request is sent to UserInfo endpoint with a holder-of-key access token.
-* A logout request is sent to Logout endpoint with a holder-of-key refresh token.
+* A logout request is sent to non-OIDC compliant {project_name} proprietary Logout endpoint with a holder-of-key refresh token.
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:88]:
```suggestion
* A logout request is sent to a non-OIDC compliant {project_name} proprietary Logout endpoint with a holder-of-key refresh token.
```

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc [hunk 5]
```diff
@@ -94,6 +94,39 @@ See https://datatracker.ietf.org/doc/html/draft-ietf-oauth-mtls-08#section-3[Mut
...
+If this setting is on, the workflow is:
+
+. A token request is sent to the token endpoint in an authorization code flow or hybrid flow.
+. {project_name} requests a DPoP proof.
+. {project_name} receives the DPoP proof.
+. {project_name} successfully verifies the DPoP proof.
+
+If verification fails, {project_name} rejects the token.
+
+In the following cases, {project_name} will verify the client sending the the refresh token:
```
Comment (andymunro) [inline docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:113]:
```suggestion
In the following cases, {project_name} will verify the client sending the refresh token:
```

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 3 above)
Comment (andymunro) [inline docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]:
```suggestion
* A token refresh request is sent to the token endpoint with a holder-of-key refresh token. This verification is done only for public clients as described in the DPoP specification. For confidential clients, the verification is not done as client authentication with proper client credentials is in place to ensure that request comes from the legitimate client. For public clients, both access tokens and refresh tokens are DPoP bound. For confidential clients, only access tokens are DPoP bound.
* A logout request is sent to a non-OIDC compliant {project_name} proprietary logout endpoint Logout endpoint with a holder-of-key refresh token. This is done only for public clients as described above.
```

**Review by andymunro — CHANGES_REQUESTED** [review]
A few changes to suggest.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 3 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]: I fixed the point.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 4 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:88]: Thank you. I fixed it.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 5 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:113]: Thank you. I fixed it.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 3 above)
Comment (tnorimat) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc:116]: Thank you. I fixed it.

**Comment by tnorimat** [comment]
@andymunro @mposolda I fixed the point you commented. Thank you.

**Review by andymunro — APPROVED** [review]
Thanks, @tnorimat Looks good.

File: docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc (hunk 1 above)
Comment (mposolda) [reply docs/documentation/server_admin/topics/clients/oidc/con-advanced-settings.adoc]: Thanks!

**Review by mposolda — DISMISSED** [review]

**Comment by tnorimat** [comment]
@mposolda Yes, I rebase and fix it.

**Review by mposolda — APPROVED** [review]

**Pull request merged**